File: ccms2/src/data/common/cms_quality.c
#include "data/common/cms_quality.h"
#include <string.h>

/* Pack 13 quality bits into a 2-byte array (big-endian) */
static void pack_quality(const cms_quality_t *q, uint8_t out[2]) {
    uint16_t bits = 0;
    bits |= (uint16_t)(q->validity.value & 0x03) << 0;
    bits |= (q->overflow.value            ? 1 : 0) << 2;
    bits |= (q->outOfRange.value          ? 1 : 0) << 3;
    bits |= (q->badReference.value        ? 1 : 0) << 4;
    bits |= (q->oscillatory.value         ? 1 : 0) << 5;
    bits |= (q->failure.value             ? 1 : 0) << 6;
    bits |= (q->oldData.value             ? 1 : 0) << 7;
    bits |= (q->inconsistent.value        ? 1 : 0) << 8;
    bits |= (q->inaccurate.value          ? 1 : 0) << 9;
    bits |= (q->substituted.value         ? 1 : 0) << 10;
    bits |= (q->test.value                ? 1 : 0) << 11;
    bits |= (q->operatorBlocked.value     ? 1 : 0) << 12;
    out[0] = (uint8_t)(bits >> 8);
    out[1] = (uint8_t)(bits);
}

static void unpack_quality(const uint8_t in[2], cms_quality_t *q) {
    uint16_t bits = ((uint16_t)in[0] << 8) | in[1];
    q->validity.value          = (bits >> 0) & 0x03;
    q->overflow.value          = (bits >> 2) & 1;
    q->outOfRange.value        = (bits >> 3) & 1;
    q->badReference.value      = (bits >> 4) & 1;
    q->oscillatory.value       = (bits >> 5) & 1;
    q->failure.value           = (bits >> 6) & 1;
    q->oldData.value           = (bits >> 7) & 1;
    q->inconsistent.value      = (bits >> 8) & 1;
    q->inaccurate.value        = (bits >> 9) & 1;
    q->substituted.value       = (bits >> 10) & 1;
    q->test.value              = (bits >> 11) & 1;
    q->operatorBlocked.value   = (bits >> 12) & 1;
}

int cms_quality_encode_stream(per_stream_t *s, const void *ptr) {
    uint8_t buf[2];
    pack_quality((const cms_quality_t*)ptr, buf);
    return cms_bit_string_fixed_encode_stream(s, buf, 13);
}

int cms_quality_decode_stream(per_stream_t *s, void *ptr) {
    uint8_t buf[2];
    int err = cms_bit_string_fixed_decode_stream(s, buf, 13);
    if (err) return CMS_ERR;
    unpack_quality(buf, (cms_quality_t*)ptr);
    return CMS_OK;
}

int cms_quality_encode(const void *ptr, uint8_t *out_buf, int *out_len) {
    per_stream_t s;
    per_stream_init_write(&s, out_buf, (size_t)*out_len);
    int rc = cms_quality_encode_stream(&s, ptr);
    if (rc) return rc;
    *out_len = (int)per_stream_bytes_written(&s);
    return CMS_OK;
}

int cms_quality_decode(void *ptr, const uint8_t *in_buf, int in_len) {
    per_stream_t s;
    per_stream_init_read(&s, in_buf, (size_t)in_len);
    return cms_quality_decode_stream(&s, ptr);
}

File: jcms/jcms-core/src/main/java/com/ysh/jcms/data/common/CmsQuality.java
package com.ysh.jcms.data.common;

import com.ysh.jcms.core.CmsType;
import com.ysh.jcms.data.InnerQuality;
import com.ysh.jcms.data.scalar.CmsBoolean;
import com.ysh.jcms.data.scalar.CmsInt32;

/**
 * Quality ::= BIT STRING (SIZE(13)) — 7.3.6
 * <p>
 * Bit 0-1 = validity (2-bit: 0=good, 1=invalid, 2=reserved, 3=questionable).
 * Bit 2-12 = single-bit boolean flags.
 */
public class CmsQuality extends CmsType {

    public CmsInt32 validity;
    public CmsBoolean overflow;
    public CmsBoolean outOfRange;
    public CmsBoolean badReference;
    public CmsBoolean oscillatory;
    public CmsBoolean failure;
    public CmsBoolean oldData;
    public CmsBoolean inconsistent;
    public CmsBoolean inaccurate;
    public CmsBoolean substituted;
    public CmsBoolean test;
    public CmsBoolean operatorBlocked;

    public CmsQuality() {
        super(new InnerQuality());
        this.validity = new CmsInt32();
        this.overflow = new CmsBoolean();
        this.outOfRange = new CmsBoolean();
        this.badReference = new CmsBoolean();
        this.oscillatory = new CmsBoolean();
        this.failure = new CmsBoolean();
        this.oldData = new CmsBoolean();
        this.inconsistent = new CmsBoolean();
        this.inaccurate = new CmsBoolean();
        this.substituted = new CmsBoolean();
        this.test = new CmsBoolean();
        this.operatorBlocked = new CmsBoolean();
    }

    public CmsQuality validity(int v) { this.validity.value(v); return this; }
    public CmsQuality overflow(boolean v) { this.overflow.value(v); return this; }
    public CmsQuality outOfRange(boolean v) { this.outOfRange.value(v); return this; }
    public CmsQuality badReference(boolean v) { this.badReference.value(v); return this; }
    public CmsQuality oscillatory(boolean v) { this.oscillatory.value(v); return this; }
    public CmsQuality failure(boolean v) { this.failure.value(v); return this; }
    public CmsQuality oldData(boolean v) { this.oldData.value(v); return this; }
    public CmsQuality inconsistent(boolean v) { this.inconsistent.value(v); return this; }
    public CmsQuality inaccurate(boolean v) { this.inaccurate.value(v); return this; }
    public CmsQuality substituted(boolean v) { this.substituted.value(v); return this; }
    public CmsQuality test(boolean v) { this.test.value(v); return this; }
    public CmsQuality operatorBlocked(boolean v) { this.operatorBlocked.value(v); return this; }

    @Override
    public void syncToInner() {
        int packed = 0;
        packed |= (validity.value() & 0x03) << 0;
        if (overflow.value()) packed |= (1 << 2);
        if (outOfRange.value()) packed |= (1 << 3);
        if (badReference.value()) packed |= (1 << 4);
        if (oscillatory.value()) packed |= (1 << 5);
        if (failure.value()) packed |= (1 << 6);
        if (oldData.value()) packed |= (1 << 7);
        if (inconsistent.value()) packed |= (1 << 8);
        if (inaccurate.value()) packed |= (1 << 9);
        if (substituted.value()) packed |= (1 << 10);
        if (test.value()) packed |= (1 << 11);
        if (operatorBlocked.value()) packed |= (1 << 12);
        ((InnerQuality) inner).value = packed;
    }

    @Override
    public void syncFromInner() {
        int packed = ((InnerQuality) inner).value;
        validity.value(packed & 0x03);
        overflow.value((packed >> 2 & 1) != 0);
        outOfRange.value((packed >> 3 & 1) != 0);
        badReference.value((packed >> 4 & 1) != 0);
        oscillatory.value((packed >> 5 & 1) != 0);
        failure.value((packed >> 6 & 1) != 0);
        oldData.value((packed >> 7 & 1) != 0);
        inconsistent.value((packed >> 8 & 1) != 0);
        inaccurate.value((packed >> 9 & 1) != 0);
        substituted.value((packed >> 10 & 1) != 0);
        test.value((packed >> 11 & 1) != 0);
        operatorBlocked.value((packed >> 12 & 1) != 0);
    }
}
